File: cpp/6.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <unordered_map>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  string convert(string s, int nRows) {
    if(nRows == 1) return s;
    vector<string> res(nRows);
    int idx = 0, d = 1;
    for(auto& c : s){
      res[idx] += c;
      idx += d;
      if(idx == -1)
        idx = 1, d = 1;
      else if(idx == nRows)
        idx = nRows-2, d = -1;
    }
    string fin;
    for(auto& r : res)
      fin += r;
    return fin;
  }
};

int main(){
  auto sol = Solution();
  cout << sol.convert("PAYPALISHIRING", 3);
}

File: python/6.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows < 2:
            return s
        res = [[] for i in xrange(numRows)]
        di = 1
        r = 0
        for c in s:
            res[r].append(c)
            if r == 0:
                di = 1
            elif r == numRows-1:
                di = -1
            r += di
        return ''.join(''.join(row) for row in res)
